### DISCUSSION TITLE
Textadept Qt, CMake, and future plans

### DISCUSSION THREAD

**Discussion by orbitalquark** [orbitalquark/textadept#300] [General]
I've been working on Qt support for Textadept in a separate branch for the past couple of weeks and now it's been merged into the default (main) branch. Going forward, Qt will be the default GUI toolkit for Textadept on Windows, macOS, and Linux. The nightlies will reflect this starting tonight.

Things to note:
* The GTK version of Textadept is still supported, but only on Linux. If you want to build with GTK, you will need to do so manually; binaries will no longer be provided\*.
* The Qt version of Textadept is not a single-instance application, only the GTK version is.
* The minimum required version of macOS is now 10.13 (High Sierra), the minimum version that Qt 5.15 supports.
* Support for GTK on Windows and macOS has been removed. (Linux is still supported.)

Textadept has switched to a CMake-based build for building natively on Windows, macOS, and Linux. The README and Manual have been updated with instructions on how to build Textadept using CMake.

The next release of Textadept will be a major one: 12.0. The 11.5 release (which was currently at the alpha stage) will be dropped. I expect the first alpha of 12.0 to arrive on 1 January 2023.

Feel free to try out the new changes, either from source or from the nightlies, and send me any feedback you might have. I am currently using the Qt version of Textadept to develop Textadept.

\* Subject to change. I am not completely sure about not providing binaries.

**Comment by orbitalquark** [comment]
Please note that at this time, if you want to compile the Qt version from source, you will need to enable `NIGHTLY` in CMake, as Scintillua and Scinterm have needed changes that are not yet in a release build.

**Comment by rgieseke** [comment]
What are the reasons for this change, easier/better with native builds on Win/macOS?
Am i supposed to see a difference on Linux (I didn't notice one)?

**Reply by orbitalquark** [reply]
The main reason for the change is to use a professionally maintained UI toolkit that looks good on all three platforms. I have been unhappy with the look and performance of Textadept on Windows and macOS ever since the switch to GTK 3 after GTK 2 was end-of-lifed. I also have zero confidence in GTK 4.

The secondary reason was to be a good citizen and allow building natively on all three platforms instead of requiring a Linux docker image to cross-compile.

A pleasant side-effect is that debugging natively compiled binaries is easier. I could never debug a cross-compiled exe on Windows, but I was able to debug a natively compiled exe with Visual Studio the other day and it was fantastic.

No, you are not supposed to see a difference. The fact that you didn't notice one on Linux means the transition is going well :)

**Reply by rgieseke** [reply]
That sounds very good :-)
Thank you for the explanations!

**Reply by ghost** [reply]
@orbitalquark You are so right to switch to Qt! The fonts quality really improved compared to the previous time I tried textadept on https://github.com/orbitalquark/textadept/issues/295 :smile:

**Reply by ghost** [reply]
Btw Gtk version should be kept for Gtk based desktop environments on Linux though. Gtk on Linux looks pretty good, but on Windows and macOS it sucks :smile:

**Comment by paaguti** [comment]
Hi :-)

I was going to ask the same... out of sheer curiosity... how does this make
the process easier in macOS?
I recognise that the macOS GTK compatibility layer is anything but
programmer-friendly

BTW, we will need to figure out (again) how to make the cmake build
packaging-friendly ;-)
I'm able to build, but can't install the link to /usr/bin/textadept because
of the way it is expressed in the CMakefile.

In my first try (running from /usr/share/textadept), I see quirks with the
hydra module too :-(

With time and patience, I'm sure we'll sort everything out. In the
meantime, could you please see into the possibility of installing file-diff
and other modules
with binary libraries into /usr/share/textadept/modules ? This would ease
the pain of generating packages for Debian and derivatives (and, I guess,
RedHat, too)

Thanks, /PA


On Sun, 4 Dec 2022 at 11:15, Robert Gieseke ***@***.***>
wrote:

> What are the reasons for this change, easier/better with native builds on
> Win/macOS?
> Am i supposed to see a difference on Linux (I didn't notice one)?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/300#discussioncomment-4302672>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4CKKATRKUKGS3RK3BLWLRVMTANCNFSM6AAAAAASTA3DZY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet

**Reply by orbitalquark** [reply]
I don't think the build process on macOS is easier. I don't think it's harder either. The end result is a better looking and more performant application on macOS and I'm quite happy with the result.

Yes, packaging is a concern. Are you able to change `CMAKE_INSTALL_PREFIX` to `/usr` and have things work? I will look into adding an install rule for binary packages, as you suggested.

Thanks for your patience. That's what this alpha stage is for.

**Reply by orbitalquark** [reply]
The file-diff, spellcheck, and debugger module have all been updated to install to `${CMAKE_INSTALL_PREFIX}/share/textadept/modules` on Linux.

**Comment by orbitalquark** [comment]
I may include the GTK version alongside the Qt version for Linux downloads. It depends on how hard it would be to maintain.

**Comment by paaguti** [comment]
I've switched to Qt everywhere and the experience is quite nice and uniform.
I don't think textadept/Qt on Linux sucks (even on GNOME or MATE).


On Sun, 18 Dec 2022 at 15:46, orbitalquark ***@***.***> wrote:

> I may include the GTK version alongside the Qt version for Linux
> downloads. It depends on how hard it would be to maintain.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/300#discussioncomment-4442139>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4H3UWHTIEQWJTME63DWN4PVNANCNFSM6AAAAAASTA3DZY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet

**Comment by mhwombat** [comment]
I’m working on updating the nix package for the upcoming release of Textadept, and have run into a hurdle.

When building a package in NixOS, it is impossible to access the Internet. Everything a package needs to build, including source code and binaries, must be obtained before the build step. This is part of the functional programming philosophy of ensuring that builds are repeatable and do not have side effects.

The previous version of Textadept was built with make. The makefile used wget to download some archives for scintilla, etc., which isn't allowed. In the Nix package, we solved this by using fetchurl to pre-fetch the archives and putting them where the makefile expected them to be. Then, when make ran, it would find the files and not try to download them.

The new version of textadept is built with CMake, and it also wants to download stuff. I tried to use the same trick of pre-fetching the archives. For example I downloaded scintilla530.tgz and put it where CMake would have put it, in build/_deps/scintilla-subbuild/scintilla-populate-prefix/src/. However, CMake tries to download it anyway. This may be because it also expects other directories such as build/_deps/scintilla-subbuild/scintilla-populate-prefix/src/scintilla-populate-stamp to be populated with a lot of additional stuff.

I’m not familiar with CMake, but a cursory web search hasn’t turned up a way (without modifying CMakeLists.txt) to tell CMake to use pre-fetched archives instead of downloading them. Do you know of a way to do this? If not, would it be possible to add an option to use pre-fetched archives during the build?

**Comment by orbitalquark** [comment]
You can place pre-fetched archives directly in ${CMAKE_BINARY_DIR}/_deps. That's what happens during the build process on GitHub in order to not download the same archives over and over again: https://github.com/orbitalquark/textadept/blob/21788cf7cca50a0acf8566beeb30d92f1903102d/.github/workflows/release.yml#L44-L45

**Reply by mhwombat** [reply]
> You can place pre-fetched archives directly in ${CMAKE_BINARY_DIR}/_deps.

That worked a treat. I've written and tested a Nix/NixOS package for the alpha version, so we'll be ready when you have the actual release.

**Comment by snoopy** [comment]
Will there be a proper dark mode for the UI going forward?

**Comment by orbitalquark** [comment]
That is one goal since I think Qt supports it. I just have to figure out how to do so :)

**Reply by ioplker** [reply]
I used a workaround while waiting* for dark mode:
1. Invert editor's color theme
2. Setup a compositor to invert whole TA windows (works for dialogs too):
```bash
picom --invert-color-include 'class_g="textadept"'
```

Nevertheless, thanks for your work @orbitalquark :+1:

\* Did search for some hours, but sadly I know nothing about Qt styling. Though have to admit that some NixOS packages don't have dark mode too.

**Comment by paaguti** [comment]
I have light and dark themes for Textadept and emacs using ThemeCreator.
AAMOF, I'm adding TextAdept12 theme support to ThemeCreator (so crossing
fingers there isn't too much more to change from here on ;-) )


On Tue, 27 Dec 2022 at 17:31, orbitalquark ***@***.***> wrote:

> That is one goal since I think Qt supports it. I just have to figure out
> how to do so :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/300#discussioncomment-4503545>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4EO3PGOBXTBYSESZ5TWPMKULANCNFSM6AAAAAASTA3DZY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet
